evolver/hardware/standard/pump.py

```diff
@@ -0,0 +1,112 @@
+                # hz values per solenoid number - containing 3 of them.
+                hz_a = [input.flow_rate] * 3
+                for solenoid, hz in enumerate(hz_a):
+                    # solenoid is 1-indexed on hardware
+                    cmd[pump * 3 + solenoid] = f"{hz}|{pump}|{solenoid+1}"
+            elif any(pump - i < 3 for i in self.ipp_pumps):
+                raise ValueError(f"pump slot {pump} reserved for IPP pump, cannot address as standard")
+            else:
+                # TODO: calibration transform here. The time_to_pump is the target variable
+                # where pump_interval would presumably be what the pump_interval was set to
+                # during calibration and is fixed post-calibration.
+                time_to_pump, pump_interval = (input.flow_rate, int(input.flow_rate))
+                cmd[pump] = f"{time_to_pump}|{pump_interval}".encode()
+        with self.serial as comm:
+            comm.communicate(SerialData(addr=self.addr, data=cmd))
+        self.committed = inputs
+
+
+class VialIEPump(EffectorDriver):
+    """Vial-based Influx-Efflux Pump driver.
+
+    Each pump in the array is either an influx (to bring liquid to the vial) or
+    efflux (to pull waste out of vial) or spare (unused). Flow rates can be configured
+    separately or as a single rate.
+
+    Configuration parameters ``influx_map`` and ``efflux_map`` can be used to specify
+    explicit mapping between vial and underlying pump_id (of :py:class:`GenericPump`),
+    which by default reserves the first 3rd of slots to influx, and second third to
+    efflux (remaining are spare).
+    """
+
+    class Config(GenericPump.Config):
+        influx_map: dict[int, int] | None = Field(None, description="map of vial to influx pump ID")
+        efflux_map: dict[int, int] | None = Field(None, description="map of vial to efflux pump ID")
+
+    class Input(VialBaseModel):
+        flow_rate_influx: float = Field(description="influx flow rate in ml/s")
+        flow_rate_efflux: float | None = Field(None, description="efflux flow rate in ml/s. Defaults to same as influx")
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        p_slots = self.slots * 3  # influx, efflux, spare (or 3 solenoids for IPP)
+        self._generic_pump = GenericPump(
+            addr=self.addr,
+            slots=p_slots,
+            serial=self.serial_conn,
+            ipp_pumps=self.ipp_pumps,
+            evolver=kwargs.get("evolver"),
+        )
+        self.influx_map = self.influx_map or {i: i for i in range(0, p_slots)}
+        self.efflux_map = self.efflux_map or {i: i + self.slots for i in range(0, p_slots)}
+
+    def commit(self):
+        for p in self.proposal.values():
+            if p.vial in self.vials:
+                if p.flow_rate_efflux is None:
+                    p.flow_rate_efflux = p.flow_rate_influx
```

Comment:
Do we need to add the alternate case so that it can be symmetric?

Should there be a flag for this behavior to allow the user to either raise or lower the fluid level?

Reply:
Will add there. I don't think it is necessary to be symmetric, it is documented that to set same rate simply either set the same rates or set only influx rate. We of course could simply require both to be explicitly set.

To raise or lower volume the user would set different rates for influx/efflux which is fully supported.

Reply:
> We of course could simply require both to be explicitly set.

I'd prefer this option for simplicity, but we'd obviously have to ask the PIs since it changes current and expected behavior.

>To raise or lower volume the user would set different rates for influx/efflux which is fully supported.

True, true, though it would be a better user experience to be able to do this directly by turning one off, rather than having to compute deltas.

Reply:
I'll make explicit required, it is not too much a burden on user. However I won't here deal with the deltas since there is some complexity as to timing which I don't think we have control over and probably have to go back to PIs to see what use case there is. Can loop back to it later